[pytorch] Add HashStore to c10d

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28921 [pytorch] Add HashStore to c10d**

This implementation is quite similar to the HashStore in gloo -
an ephemeral in-process store with a lock and unordered_map<>.

There are a few tweaks/differences based on c10d vs gloo:
  - c10d expects add/check methods
  - c10d get() use cases expect to wait up to super::timeout_ if the value isn't present
  - c10d set() isn't expected to throw if the value is present.
  - c10d uses uint8_t vs char

It's potentially a better choice for some cases than FileStore when we
don't need cross-process access, or care about the backing file.

Differential Revision: [D18233713](https://our.internmc.facebook.com/intern/diff/D18233713/)

Commit:
[pytorch] Add HashStore to c10d

This implementation is quite similar to the HashStore in gloo -
an ephemeral in-process store with a lock and unordered_map<>
There are just small tweaks for the differences (e.g. add/check methods,
uint8_t vs char)

It's potentially a better choice for some cases than FileStore when we
don't need cross-process access, or care about the backing file.

Differential Revision: [D18233713](https://our.internmc.facebook.com/intern/diff/D18233713/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Add HashStore to c10d"


This implementation is quite similar to the HashStore in gloo -
an ephemeral in-process store with a lock and unordered_map<>
There are just small tweaks for the differences (e.g. add/check methods,
uint8_t vs char)

It's potentially a better choice for some cases than FileStore when we
don't need cross-process access, or care about the backing file.

Differential Revision: [D18233713](https://our.internmc.facebook.com/intern/diff/D18233713/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Add HashStore to c10d"


This implementation is quite similar to the HashStore in gloo -
an ephemeral in-process store with a lock and unordered_map<>
There are just small tweaks for the differences (e.g. add/check methods,
uint8_t vs char)

It's potentially a better choice for some cases than FileStore when we
don't need cross-process access, or care about the backing file.

Differential Revision: [D18233713](https://our.internmc.facebook.com/intern/diff/D18233713/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Add HashStore to c10d"


This implementation is quite similar to the HashStore in gloo -
an ephemeral in-process store with a lock and unordered_map<>
There are just small tweaks for the differences (e.g. add/check methods,
uint8_t vs char)

It's potentially a better choice for some cases than FileStore when we
don't need cross-process access, or care about the backing file.

Differential Revision: [D18233713](https://our.internmc.facebook.com/intern/diff/D18233713/)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/HashStore.cpp:
```diff
@@ -0,0 +1,84 @@
+#include <stdint.h>
+#include <unistd.h>
+
+#include <errno.h>
```
Pedantic: can you group C stdlib headers and C++ stdlib headers separately?

jjlilley replied:
sure, done :)

pietern commented on torch/lib/c10d/HashStore.cpp:
```diff
@@ -0,0 +1,84 @@
+  lock.unlock();
+  wait({key}, timeout_);
+  lock.lock();
+  return map_[key];
```
`cv_.wait_for`? The `pred` can be simpler than the one used in `wait`.

jjlilley replied:
sure, done.

pietern reviewed (approved):
Thanks, @jjlilley!

Two minor nits. Feel free to merge after addressing.

Commit:
Update on "[pytorch] Add HashStore to c10d"


This implementation is quite similar to the HashStore in gloo -
an ephemeral in-process store with a lock and unordered_map<>.

There are a few tweaks/differences based on c10d vs gloo:
  - c10d expects add/check methods
  - c10d get() use cases expect to wait up to super::timeout_ if the value isn't present
  - c10d set() isn't expected to throw if the value is present.
  - c10d uses uint8_t vs char

It's potentially a better choice for some cases than FileStore when we
don't need cross-process access, or care about the backing file.

Differential Revision: [D18233713](https://our.internmc.facebook.com/intern/diff/D18233713/)

[ghstack-poisoned]

jjlilley commented:
Thanks for looking!
